kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java

```diff
@@ -238,6 +249,100 @@ public CompletionStage<Void> deregisterVirtualCluster(VirtualCluster virtualClus
         return deregisterFuture;
     }
 
+    /**
+     * Reconciles the current set of bindings for this virtual cluster against those required by the current set of nodes.
+     *
+     * @param virtualCluster virtual cluster
+     * @param nodeIds        current set of node ids
+     * @return CompletionStage yielding true if binding alterations were made, or false otherwise.
+     */
+    @Override
+    public CompletionStage<Void> reconcile(VirtualCluster virtualCluster, Set<Integer> nodeIds) {
+        Objects.requireNonNull(virtualCluster, "virtualCluster cannot be null");
+        Objects.requireNonNull(nodeIds, "nodeIds cannot be null");
+
+        var vcr = registeredVirtualClusters.get(virtualCluster);
+        if (vcr == null || vcr.deregistrationStage().get() != null) {
+            // cluster not currently registered
+            return CompletableFuture.failedStage(new IllegalStateException("virtual cluster %s not registered or is being deregistered".formatted(virtualCluster)));
+        }
+
+        // TODO: consider composing all of this after registration has finished
+
+        var rec = vcr.reconciliationEntry().get();
+        if (rec == null) {
+            // Should never happen.
+            return CompletableFuture.failedStage(new IllegalStateException("virtual cluster %s in unexpected state".formatted(virtualCluster)));
+        }
+
+        if (rec.getKey().equals(nodeIds)) {
+            // set of nodeIds already reconciled.
+            return rec.getValue();
+        }
+        else {
+            // set of nodeIds differ
+            return rec.getValue().thenCompose(u -> {
+                Map.Entry<Set<Integer>, CompletionStage<Void>> updated;
+                var cand = createReconcileEntry(nodeIds, new CompletableFuture<>());
+                if (vcr.reconciliationEntry().compareAndSet(rec, cand)) {
+                    // reconcile - work out which bindings are to be registered and which are to be removed.
+                    var bindingAddress = virtualCluster.getBindAddress().orElse(null);
+
+                    var toCreate = nodeIds.stream().map(i -> new VirtualClusterBrokerBinding(virtualCluster, i))
```

Comment:
Would this be improved by planning it out first? Collect up the set of existing VCBB, classify them into `toRemove`, `toCreate`,`toLeave` then kick off the futures?